File: app/src/main/java/com/simplecurrencyconverter/MainActivity.java
package com.simplecurrencyconverter;

import android.content.Context;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarActivity;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import com.simplecurrencyconverter.adapters.TabAdapter;
import com.simplecurrencyconverter.fragments.ConvertFragment;
import com.simplecurrencyconverter.fragments.CurrenciesFragment;
import com.simplecurrencyconverter.utils.Tab;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends ActionBarActivity
    implements CurrenciesFragment.OnFragmentInteractionListener {

    private ConvertFragment mConvertTab;

    private ViewPager mViewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        hideActionBar();

        List<Tab> tabs = new ArrayList<>();
        Fragment currenciesTab;
        if (savedInstanceState == null) {
            mConvertTab = new ConvertFragment();
            currenciesTab = new CurrenciesFragment();
        } else {
            FragmentManager fm = getSupportFragmentManager();
            List<Fragment> existingTabs = fm.getFragments();
            mConvertTab = (ConvertFragment) existingTabs.get(0);
            currenciesTab = existingTabs.get(1);
        }
        tabs.add(new Tab(getString(R.string.tab_convert), mConvertTab));
        tabs.add(new Tab(getString(R.string.tab_currencies), currenciesTab));
        addTabs(tabs);
    }

    private void hideActionBar() {
        getSupportActionBar().setTitle("");
    }

    private void addTabs(List<Tab> tabs) {
        final TabAdapter tabPagerAdapter = new TabAdapter(getSupportFragmentManager(), tabs);
        final ActionBar actionBar = getSupportActionBar();
        mViewPager = (ViewPager) findViewById(R.id.pager);

        // Add swipe views
        mViewPager.setAdapter(tabPagerAdapter);
        mViewPager.setOnPageChangeListener(
            new ViewPager.SimpleOnPageChangeListener() {
                @Override
                public void onPageSelected(int position) {
                    actionBar.setSelectedNavigationItem(position);
                }
            }
        );

        // Add tabs to action bar
        actionBar.setNavigationMode(ActionBar.NAVIGATION_MODE_TABS);
        ActionBar.TabListener tabListener = new ActionBar.TabListener() {
            @Override
            public void onTabSelected(ActionBar.Tab tab, FragmentTransaction fragmentTransaction) {
                hideKeyboard();
                mViewPager.setCurrentItem(tab.getPosition());
            }
            @Override
            public void onTabUnselected(ActionBar.Tab tab, FragmentTransaction fragmentTransaction) {
            }
            @Override
            public void onTabReselected(ActionBar.Tab tab, FragmentTransaction fragmentTransaction) {
            }
        };

        for (Tab tab : tabs) {
            actionBar.addTab(actionBar.newTab()
                .setText(tab.getTitle())
                .setTabListener(tabListener));
        }
    }

    private void hideKeyboard() {
        // From http://stackoverflow.com/a/7696791
        View view = this.getCurrentFocus();
        if (view != null) {
            InputMethodManager inputManager = (InputMethodManager) this.getSystemService(Context.INPUT_METHOD_SERVICE);
            inputManager.hideSoftInputFromWindow(view.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    @Override
    public void onCurrenciesUpdated() {
        mConvertTab.updateConversionRate();
        mViewPager.setCurrentItem(0);
    }
}

File: app/src/main/java/com/peruukki/simplecurrencyconverter/MainActivity.kt
package com.peruukki.simplecurrencyconverter

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.support.v4.view.ViewPager
import android.support.v7.app.ActionBar
import android.support.v7.app.ActionBarActivity
import android.view.inputmethod.InputMethodManager

import com.peruukki.simplecurrencyconverter.adapters.TabAdapter
import com.peruukki.simplecurrencyconverter.fragments.ConvertFragment
import com.peruukki.simplecurrencyconverter.fragments.CurrenciesFragment
import com.peruukki.simplecurrencyconverter.utils.Tab

import java.util.*

class MainActivity : ActionBarActivity(), CurrenciesFragment.OnFragmentInteractionListener {

    private var mConvertTab: ConvertFragment? = null

    private var mViewPager: ViewPager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        hideActionBar()

        val tabs = ArrayList<Tab>()
        val currenciesTab: Fragment
        if (savedInstanceState == null) {
            mConvertTab = ConvertFragment()
            currenciesTab = CurrenciesFragment()
        } else {
            val fm = supportFragmentManager
            val existingTabs = fm.fragments
            mConvertTab = existingTabs[0] as ConvertFragment
            currenciesTab = existingTabs[1]
        }
        tabs.add(Tab(getString(R.string.tab_convert), mConvertTab!!))
        tabs.add(Tab(getString(R.string.tab_currencies), currenciesTab))
        addTabs(tabs)
    }

    private fun hideActionBar() {
        supportActionBar!!.title = ""
    }

    private fun addTabs(tabs: List<Tab>) {
        val tabPagerAdapter = TabAdapter(supportFragmentManager, tabs)
        val actionBar = supportActionBar
        mViewPager = findViewById(R.id.pager) as ViewPager?

        // Add swipe views
        mViewPager!!.adapter = tabPagerAdapter
        mViewPager!!.setOnPageChangeListener(
                object : ViewPager.SimpleOnPageChangeListener() {
                    override fun onPageSelected(position: Int) {
                        actionBar!!.setSelectedNavigationItem(position)
                    }
                }
        )

        // Add tabs to action bar
        actionBar!!.navigationMode = ActionBar.NAVIGATION_MODE_TABS
        val tabListener = object : ActionBar.TabListener {
            override fun onTabSelected(tab: ActionBar.Tab, fragmentTransaction: FragmentTransaction) {
                hideKeyboard()
                mViewPager!!.currentItem = tab.position
            }

            override fun onTabUnselected(tab: ActionBar.Tab, fragmentTransaction: FragmentTransaction) {}
            override fun onTabReselected(tab: ActionBar.Tab, fragmentTransaction: FragmentTransaction) {}
        }

        for (tab in tabs) {
            actionBar.addTab(actionBar.newTab()
                    .setText(tab.title)
                    .setTabListener(tabListener))
        }
    }

    private fun hideKeyboard() {
        // From http://stackoverflow.com/a/7696791
        val view = this.currentFocus
        if (view != null) {
            val inputManager = this.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputManager.hideSoftInputFromWindow(view.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }

    override fun onCurrenciesSelected() {
        onConversionRatesUpdated()
        mViewPager!!.currentItem = 0
    }

    override fun onConversionRatesUpdated() {
        mConvertTab!!.updateConversionRate()
    }
}
